### DISCUSSION TITLE
Documentation leads me to believe all UmbracoApiControllers will just "show up" in Swagger

### DISCUSSION THREAD

**Discussion by markadrake** [umbraco/Umbraco-CMS#16895] [Features and ideas]
![image](https://github.com/user-attachments/assets/820f1703-ea47-4a07-9b11-ef47bb6ef235)

Link to Umbraco Documentation:
- [API Controllers](https://docs.umbraco.com/umbraco-cms/v/13.latest-lts/reference/routing/umbraco-api-controllers)
- [API Versioning](https://docs.umbraco.com/umbraco-cms/v/13.latest-lts/reference/api-versioning-and-openapi#adding-your-own-swagger-documents)

I followed the instructions to create my own API controllers. It's not something I often do so I prefer to have the latest and greatest documentation open at all times. I was under the impression that any `UmbracoApiController` class not mapped to a specific Swagger document would _be magically included_ on the list. But that's not so.

If you do not attribute your `UmbracoApiController` class with `[ApiController]`, and the mandatory `[Route()]`, your controller and methods will not show up in Swagger. _Maybe everyone knows this_, but I feel like the documentation should be written to a level that following the examples _will just work_.

I'm starting this as a discussion because my comfort level with C# and backend programming isn't enough to "know that I'm right".
All I do know is that Swagger did not care for my API endpoint(s) until they were annotated. 

![image](https://github.com/user-attachments/assets/9d8a7aff-4d6a-4caa-a890-efc4787f4fd4)

**Comment by nikcio** [comment]
Hey @markadrake 

I can add some insight here. Umbraco uses the `.AddApiExplorer()` and `.AddControllers` ([Code search](https://github.com/search?q=repo%3Aumbraco%2FUmbraco-CMS%20.addcontrollers&type=code)) to register what controllers and API controllers exist in the Asp Net Core application it's therefore not Umbraco directly registering what controllers are registered in the Swagger documentation. Therefore for an endpoint to show up you need to create an endpoint that will be noticed by these default Asp Net Core components.
https://github.com/umbraco/Umbraco-CMS/blob/98e9a300bc81e485eb2afc95af93c28ad3a1cea9/src/Umbraco.Web.Common/DependencyInjection/UmbracoBuilderExtensions.cs#L265

So works something like this:

Your project --> Reference to Umbraco --> Umbraco registers `.AddControllers` and `.AddApiExplorer` --> Umbraco registers swagger --> Swagger pulls the information on what endpoints exists from what is registered by `.AddControllers` and `.AddApiExplorer`. --> Swagger generates the documentation.

Therefore as you pointed out you need the `[Route("")]` attribute for it to be picked up. Now I didn't test this when writing but I think you don't nesserayly need the `[ApiController]` attribute for an endpoint to be picked up in the documentation but that just enables some other things described here: https://stackoverflow.com/questions/66545845/what-does-the-apicontroller-attribute-do

But you will also notice that if you create a minimal API endpoint [Example](https://learn.microsoft.com/en-us/aspnet/core/tutorials/min-web-api?view=aspnetcore-8.0&tabs=visual-studio#add-the-api-code) this also wouldn't be picked up in Swagger. This is because Umbraco is missing a call to `.AddEndpointsApiExplorer()` on the service collection so this will need to be added before these will show up and there also need to be made a change in the swagger options. I made a PR for this a while back but totally forgot about it: https://github.com/umbraco/Umbraco-CMS/pull/15383

That's my understanding of how it works at least hope it helps answer your question.

**Comment by c9mb** [comment]
Just starting to look at this general topic of implementing a custom REST API, and note that UmbracoApiController  was obsoleted in v14 and removed in v15.

The recommended solution is to implement a native Controller, but (and without giving it much investigation) I'm wondering if/how it might be possible to implement custom REST APIs using the DotNet Minimal API model ?
